### ISSUE TITLE
Rule Proposal: `no-unmodified-loop-condition`

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#4523]
In a loop condition, if there is a variable which is not modified in the loop body, probably it's mistake.
So this rule reports unmodified variables in a loop condition.

The following patterns are considered problems:

```js
let node = ...;
while (node) {
    doSomething(node);
}

for (let j = 0; j < xs.length; ++i) {
}
```

The following patterns are considered not problems:

```js
let node = ...;
while (node) {
    doSomething(node);

    node = node.parent;
}

for (let j = 0; j < xs.length; ++j) {
}
```

**Comment by eslintbot** [comment]
@mysticatea Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by platinumazure** [comment]
What about if you're checking an object property and passing the object to
a function that could modify it? The modification of that property would
not be detectable via static analysis, yet it may well be modified by this
other function. (Note that I'm talking specifically about property changes-
I'm well aware that reassignments to a parameter don't have any side effect
in the calling function.)

**Comment by ilyavolodin** [comment]
Agree with @platinumazure Your first example in the invalid section has a good chance of modifying `node` inside `doSomething` function (if `node` is a complex type)

**Comment by platinumazure** [comment]
Maybe it could work if the loop condition checks an identifier only?

**Comment by ilyavolodin** [comment]
I think it can be done they way @mysticatea proposes with an additional option to clear any condition variable that is passed as a parameter of a function call inside the body of the loop.

**Comment by mysticatea** [comment]
I'm thinking this rule does not support object properties in this time.
It just checks `escope.Reference` and there are write references in the loop.

**Comment by Slayer95** [comment]
It can:
```
let node = ...;
while (node) {
    doSomething(node);
}

function doSomething (n) {
node = false;
}
```

**Comment by ilyavolodin** [comment]
@Slayer95 We wouldn't be able to catch that though static analysis anyways. The example I had in mind was more like:
```js
let node = {a:10};
while (node) {
  doSomething(node);
}
function doSomething(node) {
  node = null;
}
```

**Comment by mysticatea** [comment]
@Slayer95 Ah, I see. That pattern is analyzed correctly by escope, so we can catch it with `escope.Reference ` simply.

@ilyavolodin Parameters in JS is not pointers of C :)

**Comment by nzakas** [comment]
Can you tell if the variable is modified inside the loop vs. Outside the loop?

```js
let node = ...;
while (node) {
    doSomething(node);
}

node = node.parent;
```

**Comment by mysticatea** [comment]
@nzakas There are not write references in the while-loop, so the `node` identifier at line 2 is warned. It's a cause of an infinity loop.

----

If a write reference exist in a local function, I will check whether or not the local function is called in the loop.

```js
let node = ...;
while (node) { // 4. This `node` identifier is not warned.
    doSomething(node); // 3. Called.
}

node = node.parent;

function doSomething(n) { // 2. Checks this function is called in a loop?
    node = node.parent; // 1. A write reference of node.
}
```

**Comment by nzakas** [comment]
Ah, I see. Won't this overlap with no-constant-condition?

**Comment by platinumazure** [comment]
@nzakas I wouldn't think so. `no-constant-condition` implies a condition like `true`, using a constant, but this rule would use a condition like `node` (as in, variable `node` is not falsy). The expression `node` is not a constant expression, but this rule would try to catch no modifications. Likewise, constant conditions would imply no loop variable at all, no? The only potential overlap I could see is `const` variables, since those could be adjudged "constant" for `no-constant-condition` and also non-modified variables for this proposed rule. I imagine it would be fairly easy to make a decision of which rule should cover that case.

**Comment by nzakas** [comment]
@platinumazure ah, good call

**Comment by mysticatea** [comment]
@platinumazure Thank you!

`no-constant-condition` is covering constant values in `IfStatement`s, `ConditionalExpression`s, and loop conditions.
This proposal is covering variables in loop conditions.
As @platinumazure mentioned,  only `const` variables are overlapped.

I'm OK that this is an option of `no-constant-condition`.
But I think new rule is better.

**Comment by nzakas** [comment]
Yup, sounds good!

**Comment by mysticatea** [comment]
Thank you!
I'm working on this.
